go-controller/pkg/cni/types/types.go

```diff
@@ -50,6 +50,8 @@ type NetConf struct {
 	// restart.
 	AllowPersistentIPs bool `json:"allowPersistentIPs,omitempty"`
 
+	Alias string `json:"localnetPortAlias,omitempty"`
```

Comment:
@trozet we should agree on a name to put on the configuration to allow the users to re-use an existing mapping.

Maybe this is too low level / detail. How about directly `bridgeMapping` ?

Reply:
`bridgeMapping` would make sense to me. The user flow looks something like this:

1. Admin uses `NNCP` to configure `bridge-mappings`
```
    ovn:
      bridge-mappings:
      - localnet: localnet1 
        bridge: br-ex 
        state: present 
```
2. Admin proceeds to create a `NAD` where they have to reference `localnet1` from the example above. For me, the clearest hint to what I should be setting in the `NAD` would be if it asked for `bridgeMapping`. Another option would be `localnet`, but I like it less since we already use the term for the topology

Reply:
> The part that you actually put into the NAD is the physical network name

Good point; that is **really** what I am using - the value for the localnet port `network_name` option.

I am OK with `PhysicalNetworkName`. 

@phoracek thoughts ?

Reply:
Nice, that works for me

Reply:
I am ok with the name itself. But I have some other thoughts.

While the network name is unique per network, this new field is no longer unique per network, same name can be used by multiple networks and can lead to invalid configurations. So what if there are multiple networks defined with different name but the same physical network and vlan ID? Suddenly we have to care about something that would otherwise be someone else's issue. Are we not going to care about it and let some higher orchestration layer UX take care of validating it?

From another perspective, how does OpenStack resolve this same use case? Do they have a single bridge mapping or multiple? If multiple, how does their workflow compare to ours, or what is in our workflow that makes it so cumbersome? Are we missing some orchestration UX that would automatically handle the bridge mapping to make this a non-issue? How would this feel to someone coming from OpenStack?

Both perspectives lead me to the same place: seems like a higher orchestration layer should take care of something extra. If this orchestration layer were in place, what makes the former option better than the latter?

Reply:
> So what if there are multiple networks defined with different name but the same physical network and vlan ID? 

That is completely valid scenario. Often localnet Pods won't be the only members of the VLAN. We will also have other non-Kubernetes services running there. So it is completely fine if we have NADD A and NAD B working in the same VLAN. If that is what you were concerned about.

But perhaps I did not understood what is the issue about orchestration.

Reply:
> That is completely valid scenario. Often localnet Pods won't be the only members of the VLAN. We will also have other non-Kubernetes services running there. So it is completely fine if we have NADD A and NAD B working in the same VLAN. If that is what you were concerned about.

Well I guess that is true as long as there are no overlapping IPs. So my point is, it might be fine on some cases but not in others. And now knowing if it is fine or not can become our issue.

Reply:
I see, gotcha.

The way I'm looking at localnet is "you can use this to connect VMs to network managed by YOU". For VMs, we would typically rely on DHCP and don't use OVN IPAM. I understand that with Pods IPAM would be probably more common - there we should make it clear that it is admin's responsibility to make them non-overlapping.

I believe that secondary localnet UDN will make this easier. Localnet will be configurable through CUDN only, where you have the namespace selector. So I imagine people would typically create a single CUDN for a VLAN, matching many namespaces that should have access.

Now when I think about this - introducing this new "alias" is not making this issue worse, is it? I can have two mappings on a single bridge for two different networks with overlapping IPAM...

Localnet is dangerous by definition, it should be used by admins who know what they are doing. They have to make sure there is no collision with non-Kubernetes members of the network, and between Kubernetes members.

Reply:
One of the first use cases would be to use `"PhysicalNetworkName": "physnet"`, right? That wouldn't be possible with the CUDN as is? Should this be a field in the CUDN as well? Otherwise, does this new field in the NAD make sense if the CUDN is expected to be used? I expect the CUDN to do the required validations in this case. 

In any case this answers my first question: this is a new concern for us and we plan to not really take care of it and leave it up to the admin. We know how that works for us however, we do end up dealing with it one way or another :stuck_out_tongue_winking_eye: 

> Now when I think about this - introducing this new "alias" is not making this issue worse, is it? I can have two mappings on a single bridge for two different networks with overlapping IPAM...

Right, but that wasn't ovn-k's problem because ovn-k was not handling this inter-connect between networks. Now it would be via this new Netconf API definition so we do need a clear statement on what we plan to do about it in ovn-k.

Reply:
Yeah, sharing the default network, with or without VLAN, would be very common.

Not quite sure what you mean about CUDN there. It is possible to share `physnet` using today's NAD, as long as you don't use any VLAN (@maiqueb keep me honest here). With PhysicalNetworkName the VLAN limitation would be gone. It should be also possible to reference `physnet` as any other PhysicalNetworkName from CUDN, once it gets `localnet` topology option.

> and we plan to not really take care of it and leave it up to the admin. 

+++, i like it. We should be aware of the use-case. And we should draw a clear line of "here we help you manage it, there you have to manage it yourself"

> wasn't ovn-k's problem because ovn-k was not handling this inter-connect between networks

We do support localnet since 4.14. We allowed users to connect to br-ex native network https://docs.openshift.com/container-platform/4.14/networking/multiple_networks/configuring-additional-network.html#configuring-additional-network_ovn-kubernetes-configuration-for-a-localnet-topology.  However, +1 on providing a clear statement and guidance to admins and even within OpenShift

Thank you Jaime!

Reply:
> > Not quite sure what you mean about CUDN there.
> 
> Hmm I guess I wasn't quite sure what you were referring to with CUDN either, you used the term first 😛
> 
> I was thinking on the UDN CRD, but we don't support localnet for primary networks and I am not sure we support the namespace selector for secondary networks. So not sure a consideration for the UDN CRD applies at all here.

We want to support the namespace selector for the secondary localnet UDNs for sure.

> 
> > We do support localnet since 4.14. We allowed users to connect to br-ex native network
> 
> What I mean is that effectively, inter-connecting two different ovn-k localnet networks was not possible in ovn-k. You would have to do that through the infra network which is not ovn-k responsibility. Now with this `PhysicalNetworkName`, kinda makes that interconnect possible (at least semantically in the ovn-k netconf API) through ovn-k so it would be reasonable to think that any sanity checks for that fall under ovn-k responsibility.

I don't follow: you can have multiple networks in the same OVS bridge *today* with the localnet feature. Each network will use a different patch port. 

All this PR does is allow these multiple networks to reach the bridge (still using different patch ports) using localnet LSPs whose network name is the same. 

The admin is still responsible to ensure the physical network configuration makes sense. 

Maybe a couple of examples of whatever sanity checks you have in mind would make this clearer for me.

Reply:
> I don't follow: you can have multiple networks in the same OVS bridge _today_ with the localnet feature. Each network will use a different patch port.

I am not explaining myself very well. I know you can do that today. I am not saying you can't.

But doing so today requires two bridge mappings `physnet:br-ex,localnet:br-ex`. OVN-K does not care, is not aware nor handles the bridge mappings, so if this leads to any problem it is not OVN-K responsibility. `Something else` is setting the bridge mappings and that `something else` is responsible.

Now with this new attribute `PhysicalNetworkName` as part of the OVN-K netconf API and set to `physnet` you only need the default bridge mapping `physnet:br-ex` to have the equivalent functionality as before. In the general case this attribute makes possible to map multiple networks to the same bridge, **through OVN-K API**, introducing a gray area of responsibility whereas before it was clearly not OVN-K responsibility because OVN-K was simply not aware.

One example of the checks we could potentially make:
Check that all the localnets with the same `PhysicalNetworkName` and VlanID don't have overlapping IPs.

> We want to support the namespace selector for the secondary localnet UDNs for sure.

So the plan is also to have this field in the CRD as well?

Reply:
> > We want to support the namespace selector for the secondary localnet UDNs for sure.

> So the plan is also to have this field in the CRD as well?

Yes.

> Now with this new attribute PhysicalNetworkName as part of the OVN-K netconf API and set to physnet you only need the default bridge mapping physnet:br-ex to have the equivalent functionality as before. In the general case this attribute makes possible to map multiple networks to the same bridge, through OVN-K API, introducing a gray area of responsibility whereas before it was clearly not OVN-K responsibility because OVN-K was simply not aware.

I understand better what you mean. 

IMHO, we can simply call out in the docs this is *not* a gray area of responsibility, and that OVN-K is in no way responsible for validating the physical network configuration - the admin is.

> One example of the checks we could potentially make:
Check that all the localnets with the same PhysicalNetworkName and VlanID don't have overlapping IPs.

Thanks for the example; I just don't see how can we perform or enforce such validation: localnet networks (at least for virt use cases) are commonly known for not having subnets defined. I.e. ovn-k has no clue of the IP assignments of the workloads in it.

While saying that, I do see the potential for an higher orchestration layer you've called out in previous comments.

Reply:
> IMHO, we can simply call out in the docs this is _not_ a gray area of responsibility, and that OVN-K is in no way responsible for validating the physical network configuration - the admin is.

That works for the NAD but I don't think we can do that for the CRD because validation is one of its main purposes.

> I just don't see how can we perform or enforce such validation: localnet networks (at least for virt use cases) are commonly known for not having subnets defined.

For the ones that have it defined I guess. We have the responsibility to think about all use cases not just virt.

Reply:
> > IMHO, we can simply call out in the docs this is _not_ a gray area of responsibility, and that OVN-K is in no way responsible for validating the physical network configuration - the admin is.
> 
> That works for the NAD but I don't think we can do that for the CRD because validation is one of its main purposes.

Sure, let's defer validation to the cUDN CRD then. That makes sense to me.

> 
> > I just don't see how can we perform or enforce such validation: localnet networks (at least for virt use cases) are commonly known for not having subnets defined.
> 
> For the ones that have it defined I guess. We have the responsibility to think about all use cases not just virt.

I'm OK with best effort - i.e. validating whenever possible.

Reply:
Renamed the attribute to `PhysicalNetworkName`.

Please take another look @trozet